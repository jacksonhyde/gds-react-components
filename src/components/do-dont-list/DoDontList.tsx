import React, { HTMLProps, createContext, useContext } from 'react';
import classNames from 'classnames';
import { Tick, Cross } from '../icons';
import HeadingLevel, { HeadingLevelType } from '../../util/HeadingLevel';

type ListType = 'do' | 'dont';

interface DoDontListProps extends HTMLProps<HTMLDivElement> {
  listType: ListType;
  heading?: string;
  headingLevel?: HeadingLevelType;
}

interface DoDontList extends React.FC<DoDontListProps> {
  Item: React.FC<DoDontItemProps>;
}

const DoDontListContext = createContext<ListType>('do');

const DoDontList: DoDontList = ({
  className,
  listType,
  children,
  heading,
  headingLevel,
  ...rest
}) => (
  <div className={classNames('govuk-do-dont-list', className)} {...rest}>
    <HeadingLevel className="govuk-do-dont-list__label" headingLevel={headingLevel}>
      {heading || (listType === 'do' ? 'Do' : "Don't")}
    </HeadingLevel>
    <ul
      className={classNames(
        'govuk-list',
        { 'govuk-list--tick': listType === 'do' },
        { 'govuk-list--cross': listType === 'dont' },
      )}
    >
      <DoDontListContext.Provider value={listType}>{children}</DoDontListContext.Provider>
    </ul>
  </div>
);

interface DoDontItemProps extends HTMLProps<HTMLLIElement> {
  listItemType?: ListType;
}

const DoDontItem: React.FC<DoDontItemProps> = ({ children, listItemType, ...rest }) => {
  const listItem = useContext(DoDontListContext);
  return (
    <li {...rest}>
      {(listItemType || listItem) === 'do' ? <Tick /> : <Cross />}
      {children}
    </li>
  );
};

DoDontList.Item = DoDontItem;

export default DoDontList;
