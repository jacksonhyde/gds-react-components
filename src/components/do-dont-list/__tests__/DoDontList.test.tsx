import React from 'react';
import { shallow, mount } from 'enzyme';
import DoDontList from '..';

describe('DoDontList', () => {
  it('matches snapshot', () => {
    const doElement = shallow(<DoDontList listType="do" />);
    const dontElement = shallow(<DoDontList listType="dont" />);
    expect(doElement).toMatchSnapshot('DoDontList-Do');
    expect(dontElement).toMatchSnapshot('DoDontList-Dont');
    doElement.unmount();
    dontElement.unmount();
  });

  it('adds the correct header', () => {
    const doElement = mount(<DoDontList listType="do" />);
    const dontElement = mount(<DoDontList listType="dont" />);
    expect(doElement.find('h3.govuk-do-dont-list__label').text()).toEqual('Do');
    expect(dontElement.find('h3.govuk-do-dont-list__label').text()).toEqual("Don't");
    doElement.unmount();
    dontElement.unmount();
  });

  it('adds the correct classes', () => {
    const doElement = mount(<DoDontList listType="do" />);
    const dontElement = mount(<DoDontList listType="dont" />);

    expect(doElement.find('ul').hasClass('govuk-list--tick')).toBeTruthy();
    expect(doElement.find('ul').hasClass('govuk-list--cross')).toBeFalsy();

    expect(dontElement.find('ul').hasClass('govuk-list--tick')).toBeFalsy();
    expect(dontElement.find('ul').hasClass('govuk-list--cross')).toBeTruthy();

    doElement.unmount();
    dontElement.unmount();
  });

  describe('DoDontList.Item', () => {
    it('matches snapshot', () => {
      const element = shallow(<DoDontList.Item>Text</DoDontList.Item>);
      expect(element).toMatchSnapshot('DoDontList.Item');
      element.unmount();
    });

    it('inherits type from context', () => {
      const doList = mount(
        <DoDontList listType="do">
          <DoDontList.Item>Do</DoDontList.Item>
        </DoDontList>,
      );
      const dontList = mount(
        <DoDontList listType="dont">
          <DoDontList.Item>Don&apos;t</DoDontList.Item>
        </DoDontList>,
      );

      expect(doList.find('.govuk-icon__tick').exists()).toBeTruthy();
      expect(doList.find('.govuk-icon__cross').exists()).toBeFalsy();

      expect(dontList.find('.govuk-icon__tick').exists()).toBeFalsy();
      expect(dontList.find('.govuk-icon__cross').exists()).toBeTruthy();

      doList.unmount();
      dontList.unmount();
    });
  });
});
