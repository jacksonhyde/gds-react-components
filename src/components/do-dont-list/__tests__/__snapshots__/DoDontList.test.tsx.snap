// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoDontList DoDontList.Item matches snapshot: DoDontList.Item 1`] = `
<li>
  <Tick />
  Text
</li>
`;

exports[`DoDontList matches snapshot: DoDontList-Do 1`] = `
<div
  className="govuk-do-dont-list"
>
  <HeadingLevel
    className="govuk-do-dont-list__label"
    headingLevel="h3"
  >
    Do
  </HeadingLevel>
  <ul
    className="govuk-list govuk-list--tick"
  >
    <ContextProvider
      value="do"
    />
  </ul>
</div>
`;

exports[`DoDontList matches snapshot: DoDontList-Dont 1`] = `
<div
  className="govuk-do-dont-list"
>
  <HeadingLevel
    className="govuk-do-dont-list__label"
    headingLevel="h3"
  >
    Don't
  </HeadingLevel>
  <ul
    className="govuk-list govuk-list--cross"
  >
    <ContextProvider
      value="dont"
    />
  </ul>
</div>
`;
